### PR TITLE
add language features for .pio_version 1

### LINKED ISSUE #67: support `.pio_version 1`
and the new instructions in the RP2350 variant of the PIO peripheral

### PR DESCRIPTION
**Pull request by jepler**
New StateMachine constructor parameters will need to be added in the core. However, when these features are not used, the new constructor parameters will not be in `pio_kwargs` so backward compatibility is preserved.

Directives:
- [x] pio_version
- [x] fifo
- [x] mov_status
- [x] in
- [x] out
- [x] set

Instructions:

- [x] mov rxfifo[]
- [x] mov pindirs
- [x] irq next/prev
- [x] wait jmppin
- [x] wait irq next/prev

Closes #67

### CODE REVIEW

**Comment by jepler** [comment]
I think all the new directives & instructions are handled. However, I'd like to wait to merge this until the new StateMachine constructor parameters are supported in the core, just in case some changes end up needed there.

File: adafruit_pioasm.py [hunk 1]
```diff
@@ -58,18 +68,65 @@ def __init__(self, text_program: str, *, build_debuginfo: bool = False) -> None:
...
+                raise RuntimeError(
+                    f"{instruction} requires .pio_version {required_version}"
+                )
+
+        def int_in_range(arg, low, high, what, radix=0):
+            result = int(arg, radix)
+            if low <= result < high:
+                return result
+            raise RuntimeError(
+                f"{what} must be at least {low} and no greater than {high}, got {result}"
```
Comment (tannewt) [inline adafruit_pioasm.py:100]:
```suggestion
                f"{what} must be at least {low} and less than {high}, got {result}"
```

File: adafruit_pioasm.py [hunk 2]
```diff
@@ -79,6 +136,89 @@ def __init__(self, text_program: str, *, build_debuginfo: bool = False) -> None:
...
+                    else:
+                        mov_status_n = 0
+                    if words[idx] != "set":
+                        raise RuntimeError(f"Invalid {line})")
+                    mov_status_n |= int_in_range(words[idx + 1], 0, 8, "mov_status irq")
+                require_version(required_version, line)
+            elif words[0] == ".out":
+                require_before_instruction()
+                out_count = int_in_range(
+                    words[1], 32 if pio_version == 0 else 1, 33, ".out count"
```
Comment (tannewt) [inline adafruit_pioasm.py:173]: Does this mean 32 is the only out_count valid for version 0? It looks like this is a copy/paste issue from `.in`.

File: adafruit_pioasm.py [hunk 3]
```diff
@@ -138,16 +287,36 @@ def __init__(self, text_program: str, *, build_debuginfo: bool = False) -> None:
...
-                num = int(instruction[3], 0)
-                if not 0 <= num <= 31:
-                    raise RuntimeError("Wait num out of range")
-                assembled[-1] |= num
-                if instruction[-1] == "rel":
-                    assembled[-1] |= 0x10  # Set the high bit of the irq value
+                if instruction[2] == "jmppin":
+                    require_version(1, "wait jmppin")
+                    num = 0
+                    print("wait jmppin", instruction)
```
Comment (tannewt) [inline adafruit_pioasm.py:297]:
```suggestion
```

File: adafruit_pioasm.py [hunk 4]
```diff
@@ -58,18 +68,65 @@ def __init__(self, text_program: str, *, build_debuginfo: bool = False) -> None:
...
+            require_version(1, line)
+            if (  # pylint: disable=consider-using-in
+                fifo_type != "putget" and fifo_type != fifo_dir
+            ):
+                raise RuntimeError(
+                    f"FIFO must be configured for '{fifo_dir}' or 'putget' for {line}"
+                )
+            if arg.endswith("[y]"):
+                return 0b1000
+            return int_in_range(arg[7:-1], 0, 8, "rxfifo index")
```
Comment (tannewt) [inline adafruit_pioasm.py:113]: I think this is inverted. bit 3 is high when the index is in the instruction. So y is 0b0000.

The included index is only 2 bit too so the range is 0, 4.

File: tests/test_version.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+def test_dot_out() -> None:
+    assert_pio_kwargs(
+        ".out 32 left auto 11",
+        sideset_enable=0,
+        auto_pull=True,
+        pull_threshold=11,
+        out_shift_right=False,
+    )
+    assert_assembly_fails(".out 16")
```
Comment (tannewt) [inline tests/test_version.py:98]: This should pass.

**Review by tannewt — CHANGES_REQUESTED** [review]
Thanks for taking this on! There are a lot of little details and I think I found a couple spots that don't match the datasheet.

File: adafruit_pioasm.py (hunk 2 above)
Comment (jepler) [reply adafruit_pioasm.py]: yup I think it was incorrect.

**Comment by jepler** [comment]
oops, re-requested review but I missed a couple of points. will continue fixing.

File: adafruit_pioasm.py (hunk 4 above)
Comment (jepler) [reply adafruit_pioasm.py:113]: Yes, you were right.

I added to the tests ~1200 instructions, "substantially all" of them, generated by script invoking sdk pioasm. Now all 1200 of them are parsed by adafruit_pioasm and give the same instruction word (this does not test all side set & delay values though)

Given this bug it's a bit of a surprise that my rxfifo-using program worked, I wonder why/how. It was using an invalid `mov ,rxfifo` instruction! interesting.

**Comment by jepler** [comment]
OK now it's actually read for re-review (though I'm not likely to act on anything today or probably before monday so no rush)

File: tests/make_all.py
Comment (tannewt) [inline tests/make_all.py:1]: Please move this to `../tools` so that it is clear it isn't a test and reference it in the two related files.

**Review by tannewt — CHANGES_REQUESTED** [review]
One structural request. Good otherwise. Thanks for generating tests for everything!

File: tests/make_all.py
Comment (jepler) [reply tests/make_all.py:1]: Just noticed this. Done!

**Review by tannewt — APPROVED** [review]
Thank you!

**Pull request merged**